text/0002-extensible-provisioning-templates.md

```diff
@@ -0,0 +1,118 @@
+-   Feature Name: Extensible provisioning templates
+-   Start Date: 2016-06-28
+-   RFC PR: (leave this empty)
+-   Issue: (leave this empty)
+
+# Summary
+
+[summary]: #summary
+
+This is a suggestion for refactoring Foreman's template generation mechanism is such a way that will enable
+different plugins to add plugin-specific code to the rendered template.
+
+# Motivation
+
+[motivation]: #motivation
+
+## Original use case
+
+As part of the effort to clean up Foreman's core from puppet related code,
+I have stumbled upon `Host::Managed#handle_ca` method. This method is responsible
+for dealing with certificate autosigning mechanism. The method is called from different parts
+in the code: from `UnattendedController` and from two orchestration concerns: `Orchestration::SSHProvision`
+and `Orchestration::Compute`. The motivation for this method is to keep the autosigning window open
+for as shortest period of time as possible.
+My suggestion is to solve this design flaw by using extensible templates in the following process:
+
+1.  Foreman will generate a certificate during the regular orchestration process and store it as part of
```

Comment:
I guess historic reasons?

I'm not sure if this proposal actually simplifies the handle_ca… As you would not need also to solve generating of the keys on the server (or orchestrating the call to CA authority to generate the certs/keys) and storing it somewhere as well on the server. When I have two solutions, where one stores private keys on server and the other one not, guess what my winner is :)

Also, I'm not sure the security is at the same level here: with the autosign case, once the attacker gets his certificate issued before the provisioned host, you will at least get an error on the provisioned host once it tries to get it's certs signed. With the cert-inilne case, the attacker can gain the cert without any error to appear anywhere.

I think the secured storage deserves a RFC on it's own, the generation of certs (if we wanted to go that path) another one.

I think the extendable templates feature makes a ton of sense regardless on how we handle the certs. Would it makes sense to make two RFCs from this and discuss separately the certs part and the templates part?